src/WinRT.Runtime2/InteropServices/RestrictedErrorInfo.cs

```diff
@@ -62,26 +294,155 @@ public static void ThrowExceptionForHR(HRESULT errorCode)
     /// <seealso cref="Marshal.GetExceptionForHR(int)"/>
     public static HRESULT GetHRForException(Exception? exception)
     {
-        // TODO
-        return 0;
+        if (exception == null)
+        {
+            throw new ArgumentNullException("exception");
+        }
+        int hr = exception.HResult;
+        try
+        {
+            if (exception.TryGetRestrictedLanguageErrorInfo(out WindowsRuntimeObjectReference? restrictedErrorObject, out bool _))
+            {
+                if (restrictedErrorObject != null)
+                {
+                    IRestrictedErrorInfoMethods.GetErrorDetails(restrictedErrorObject.GetThisPtr(), out hr);
+                    GC.KeepAlive(restrictedErrorObject);
```

Comment:
Calling `GetThisPtr()` will leak the pointer here, as we never release it. This part needs some tweaks compared to CsWinRT 2.x, as we changed a lot of the reference tracking logic here. Basically what you can do here is to go through a value, which will get a "reference tracker lease" for the object reference, and pass a pointer from there. It also has the benefit that can entirely skip the native `AddRef` and `Release` pair of calls. Do this:
```csharp
using WindowsRuntimeObjectReferenceValue restrictedErrorObjectValue = restrictedErrorObject.AsValue();
```
Then pass `GetThisPtrUnsafe()` from that value to that `GetErrorDetails` call.
Also don't need `GC.KeepAlive` anymore, the value will take care of things 🙂